### PR TITLE
move frame related functions to common utils folder

### PR DESCRIPTION
**Pull request by Pnom**

### CODE REVIEW

File: src/analog/commands/index.ts [hunk 1]
```diff
@@ -4,5 +4,19 @@
...
  */
 
+import {
+    downlinkCommandById,
+    requestByResponse,
+    responseByRequest,
+    uplinkCommandById
+} from '../constants/commandRelations.js';
+
+export {
```
Comment (DarkPark) [inline src/analog/commands/index.ts:14]: this file should export only commands

File: src/analog/constants/commandRelations.ts [hunk 2]
```diff
@@ -3,16 +3,18 @@
 import * as downlinkCommands from '../commands/downlink/index.js';
 import * as uplinkCommands from '../commands/uplink/index.js';
 
+export const downlinkCommandById = Object.fromEntries(
```
Comment (DarkPark) [inline src/analog/constants/commandRelations.ts:6]: do we really need this change?

File: src/mtx/message.ts [hunk 3]
```diff
@@ -12,7 +12,8 @@ import getBytesFromBase64 from '../utils/getBytesFromBase64.js';
 import getHexFromBytes from '../utils/getHexFromBytes.js';
 import getBase64FromBytes from '../utils/getBase64FromBytes.js';
 import mergeUint8Arrays from '../utils/mergeUint8Arrays.js';
-import {aes, arrayStuff, arrayUnstuff, calculateCrcBytes} from './utils/crypto.js';
+import {aes, calculateCrcBytes} from './utils/crypto.js';
```
Comment (DarkPark) [inline src/mtx/message.ts:15]: how about use the new calculateCrc16 function?

File: src/utils/stringCollector.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+export default class StringCollector {
+    private response: string = '';
+
+    process ( data: AllowSharedBufferSource ): Array<string> {
+        const decoder = new TextDecoder();
+
+        this.response += decoder.decode(data);
+
+        if ( this.response.length !== 0) {
+            const lastLineValid = this.response.endsWith('\r\n');
```
Comment (DarkPark) [inline src/utils/stringCollector.ts:10]:
```suggestion
            const isLastLineValid = this.response.endsWith('\r\n');
```

or even `isValidEnd` as we have only one line at this moment

File: src/utils/stringCollector.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+export default class StringCollector {
```
Comment (DarkPark) [inline src/utils/stringCollector.ts:1]: can't find this class uses

do we need it? if yes - do we need this strictly as class? why not a sinle method?

Comment (DarkPark) [inline src/utils/stringCollector.ts:1]: some description with example would be nice

File: src/utils/getHexFromBytes.ts [hunk 6]
```diff
@@ -11,7 +11,7 @@ import {hexFormatOptions, IHexFormatOptions} from '../config.js';
  * input: [2, 5, 12, 255, 105, 139, 125]
  * output: '02 05 0c ff 69 8b 7d' or '0x02 0x05 0x0c 0xff 0x69 0x8b 0x7d'
  */
-export default ( buffer: Uint8Array, options: IHexFormatOptions = {} ): string => {
+export default ( buffer: Uint8Array | Array<number>, options: IHexFormatOptions = {} ): string => {
```
Comment (DarkPark) [inline src/utils/getHexFromBytes.ts:14]: the idea seems valid but the result may be unpredictable as now you can pass `[1, 2, 333333]` in this function

File: src/utils/byteStuffing.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    return byte;
+};
+
+export const arrayStuff = ( data: Array<number> ): Array<number> => {
+    const result: Array<number> = [];
+
+    data.forEach(byte => {
+        const stuff = byteStuff(byte);
```
Comment (DarkPark) [inline src/utils/byteStuffing.ts:33]: how about this?

```js
const byteStuffMap = {0x13: 0x33, 0x11: 0x31, 0x7d: 0x5d, 0x7e: 0x5e};

const stuff = byteStuffMap[byte] || byte;
```

File: src/analog/message.ts [hunk 8]
```diff
@@ -144,6 +145,30 @@ export const fromBytes = ( data: Uint8Array, config?: IMessageConfig ) => {
...
+export const fromFrames = ( frames: Array<Array<number> | Uint8Array>, config?: IMessageConfig ) => {
+    const commands = [];
+
+    for ( const frame of frames ) {
+        const commandsInFrame = fromFrame(frame, config);
+
+        commands.push(...commandsInFrame);
+    }
+
+    return commands.flatMap( command => command.command);
```
Comment (DarkPark) [inline src/analog/message.ts:169]:
```suggestion
    return commands.flatMap(command => command.command);
```

File: src/utils/frame.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        return null;
+    }
+
+    const crcBuffer = new BinaryBuffer(new Uint8Array(frame.slice(-2)));
+
+    return crcBuffer.getUint16();
+};
+
+
+export const toFrame = ( content: Array<number> | Uint8Array ): Array<number> => {
```
Comment (DarkPark) [inline src/utils/frame.ts:26]: ): Uint8Array

?

File: src/analog/message.ts [hunk 10]
```diff
@@ -144,6 +145,30 @@ export const fromBytes = ( data: Uint8Array, config?: IMessageConfig ) => {
     return result;
 };
 
+export const fromFrame = ( frame: Array<number> | Uint8Array, config?: IMessageConfig ) => {
```
Comment (DarkPark) [inline src/analog/message.ts:148]: would be nice to do this for mtx

File: src/utils/frameCollector.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
+import {START_BYTE, STOP_BYTE} from '../constants/frameAttributes.js';
+
+
+class FrameCollector {
+    private frames: Array< Array<number> > = [];
+
+    private frame: Array<number> = [];
+
+    private startByteReceived: boolean = false;
```
Comment (DarkPark) [inline src/utils/frameCollector.ts:9]:
```suggestion
    private isStartByteReceived: boolean = false;
```

File: src/utils/byteStuffing.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        case 0x5D: return 0x7D;
+        case 0x5E: return 0x7E;
+        default:
+            break;
+    }
+
+    return byte;
+};
+
+export const arrayStuff = ( data: Array<number> ): Array<number> => {
```
Comment (DarkPark) [inline src/utils/byteStuffing.ts:29]: Uint8Array ?

File: src/analog/message.ts [hunk 13]
```diff
@@ -162,3 +187,9 @@ export const toBytes = ( commands: Array<Command> ): Uint8Array => {
 export const toHex = ( commands: Array<Command>, options: IHexFormatOptions = {} ): string => getHexFromBytes(toBytes(commands), options);
 
 export const toBase64 = ( commands: Array<Command> ): string => getBase64FromBytes(toBytes(commands));
+
+export const toFrame = ( commands: Array<Command> ): Uint8Array => {
+    const data = toBytes(commands);
+
+    return new Uint8Array(Frame.toFrame(data));
```
Comment (DarkPark) [inline src/analog/message.ts:194]:
```suggestion
    return new Uint8Array(Frame.toFrame(toBytes(commands)));
```

?

File: src/analog/commands/index.ts (hunk 1 above)
Comment (Pnom) [reply src/analog/commands/index.ts:14]: fixed

File: src/analog/constants/commandRelations.ts (hunk 2 above)
Comment (Pnom) [reply src/analog/constants/commandRelations.ts:6]: downlinkCommandById in commandRelations.ts now

File: src/mtx/message.ts (hunk 3 above)
Comment (Pnom) [reply src/mtx/message.ts:15]: calculateCrcBytes implemented with calculateCrc16

File: src/utils/getHexFromBytes.ts (hunk 6 above)
Comment (Pnom) [reply src/utils/getHexFromBytes.ts:14]: fixed. use Uint8Array everywhere

File: src/utils/stringCollector.ts (hunk 4 above)
Comment (Pnom) [reply src/utils/stringCollector.ts]: String collector removed from the request

File: src/utils/stringCollector.ts (hunk 5 above)
Comment (Pnom) [reply src/utils/stringCollector.ts]: String collector removed from the request. Will be added later with tests

Comment (Pnom) [reply src/utils/stringCollector.ts]: String collector removed from the request. Will be added later with tests

File: src/utils/byteStuffing.ts (hunk 7 above)
Comment (Pnom) [reply src/utils/byteStuffing.ts]: fixed

File: src/analog/message.ts (hunk 8 above)
Comment (Pnom) [reply src/analog/message.ts]: fixed

File: src/utils/frame.ts (hunk 9 above)
Comment (Pnom) [reply src/utils/frame.ts]: fixed

File: src/analog/message.ts (hunk 10 above)
Comment (Pnom) [reply src/analog/message.ts]: It will be. too much work for one merge

File: src/utils/byteStuffing.ts (hunk 12 above)
Comment (Pnom) [reply src/utils/byteStuffing.ts]: fixed

File: src/analog/message.ts (hunk 8 above)
Comment (Pnom) [reply src/analog/message.ts]: Changes in analog/message.ts are removed from the request for now.

File: src/utils/frameCollector.ts (hunk 11 above)
Comment (Pnom) [reply src/utils/frameCollector.ts]: fixed

File: src/utils/frame.ts [hunk 14]
```diff
@@ -0,0 +1,128 @@
+import calculateCrc16 from './calculateCrc16.js';
+import getBytesFromHex from './getBytesFromHex.js';
+import getBytesFromBase64 from './getBytesFromBase64.js';
+import BinaryBuffer from './BinaryBuffer.js';
+
+
+const STUFFING_BYTE = 0x7d;
+
+const byteStuff = ( byte: number ): number => {
+    const byteStuffMap: Record<number, number> = {0x13: 0x33, 0x11: 0x31, 0x7d: 0x5d, 0x7e: 0x5e};
```
Comment (DarkPark) [inline src/utils/frame.ts:10]: it's better to declare it outside this function as now this is recreating on each byteStuff call

File: src/utils/frame.ts [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+
+    return byteStuffMap[byte] || byte;
+};
+
+const byteUnstuff = ( byte: number ): number => {
+    const byteUnstuffMap: Record<number, number> = {0x33: 0x13, 0x31: 0x11, 0x5d: 0x7d, 0x5e: 0x7e};
```
Comment (DarkPark) [inline src/utils/frame.ts:16]: the same

and the function body is very small now

File: tests/utils/frame.collector.test.ts [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+            '7e ff ff ff'
+        ];
+        const frames = collectFrames(sequence);
+
+        expect(frames.length).toBe(0);
+    });
+
+    test('test sequence with small frame', () => {
+        const sequence = [
+
```
Comment (DarkPark) [inline tests/utils/frame.collector.test.ts:45]:
```suggestion
```

File: tests/utils/frame.collector.test.ts [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+import getBytesFromHex from '../../src/utils/getBytesFromHex.js';
+
+
+describe('frame collector tests', () => {
+    const collectFrames = ( hexArray: Array<string> ): Array<IFrame> => {
+        const frameCollector = new FrameCollector();
+        let frames: Array<IFrame> = [];
+
+        for ( let index = 0; index < hexArray.length; index++ ) {
+            frames = [
```
Comment (DarkPark) [inline tests/utils/frame.collector.test.ts:12]:
```suggestion
            frames = frames.concat(frameCollector.process(getBytesFromHex(hexArray[index])));
```

File: src/utils/frame.ts [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+            hasPrevStuffingByte = false;
+            result[position++] = byteUnstuff(byte);
+        } else {
+            result[position++] = byte;
+        }
+    });
+
+    return new Uint8Array(result);
+};
+
```
Comment (DarkPark) [inline src/utils/frame.ts:58]:
```suggestion
```

File: src/analog/constants/commandRelations.ts [hunk 19]
```diff
@@ -9,8 +9,8 @@ export const requestByResponse = new Map();
 export const responseByRequest = new Map();
 
 // id to constructor
-export const requestById = new Map();
-export const responseById = new Map();
+export const requestById = new Map(Object.values(downlinkCommands).map(item => [item.id, item]));
```
Comment (DarkPark) [inline src/analog/constants/commandRelations.ts:12]:
```suggestion
export const requestById = new Map(Object.values(downlinkCommands).map(command => [command.id, command]));
```

?

File: src/utils/frame.ts (hunk 14 above)
Comment (Pnom) [reply src/utils/frame.ts]: fixed

File: src/utils/frame.ts (hunk 15 above)
Comment (Pnom) [reply src/utils/frame.ts]: fixed

File: tests/utils/frame.collector.test.ts (hunk 16 above)
Comment (Pnom) [reply tests/utils/frame.collector.test.ts]: fixed

File: tests/utils/frame.collector.test.ts (hunk 17 above)
Comment (Pnom) [reply tests/utils/frame.collector.test.ts]: fixed

File: src/utils/frame.ts (hunk 18 above)
Comment (Pnom) [reply src/utils/frame.ts:52]: fixed

File: src/analog/constants/commandRelations.ts (hunk 19 above)
Comment (Pnom) [reply src/analog/constants/commandRelations.ts]: fixed

**Review by DarkPark — APPROVED** [review]

**Review by DeadByElpy — APPROVED** [review]

**Pull request merged**